### DISCUSSION TITLE
SessionMiddleware creates warnings on modern browsers if used on an endpoint for a site hosted in another site's iframe

### DISCUSSION THREAD

**Discussion by owenlamont** [encode/starlette#2441] [Potential Issue]
I was using SessionMiddleware in an endpoint for a site whose page was embedded in an iframe and after a few attempts to get the parameters right (I realised I need to set _same_site_ to "none" I got it to work - but Chrome (I'm currently using version 20) was still spamming warnings at me saying this approach would soon be blocked by third party cookie restrictions in the future [and referring me to this help page](https://developers.google.com/privacy-sandbox/3pcd#report-issues).

For my use case the [CHIPS/partitioned cookies](https://developers.google.com/privacy-sandbox/3pcd/chips) seemed to be the right approach as the domain for the parent iframe was fixed - but that's where I ran into an issue with SessionMiddleware because I couldn't specify the _Partitioned_ security option needed to enable CHIPS.

I'm currently using Starlette 0.35.1 - although I believe this is still an issue on current master branch.

The main issue is not being able to pass the ["partitioned" parameter to the Set-Cookie call](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#partitioned_cookie).

Some secondary minor issues are the same_site = "none" value actually constrains several other set cookie parameter values that SessionMiddleware doesn't enforce, namely:

- same_site = "none" requires https so the https_only must also be set to true
- same_site = "none" if used with partitioned requires domain-bound cookie names - so the cookie name (stored in session_cookie) requires a __Host- prefix
- A secondary consequence of setting the __Host- prefix in the cookie name is [the path must remain as the default "/"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#cookie_prefixes)

To work around these problems I created my own copy of SessionMiddleware to use for now (although I'd love to go back to using the Starlette implementation if this can be resolved here). The last 5 lines of the dunder init are the main changes. 

Note in this implementation I didn't add any logic to enforce the __Host- cookie name prefix or block the path from being changed. I just relied on the calling code to set that properly. With this new implementation and keeping the cookie name and path valid I was able to get (this version of) SessionMiddleware to work for a site in an iframe without triggering any warnings in Chrome.

```python
from base64 import b64decode, b64encode
import json
from typing import Literal

import itsdangerous
from itsdangerous.exc import BadSignature
from starlette.datastructures import MutableHeaders, Secret
from starlette.requests import HTTPConnection
from starlette.types import ASGIApp, Message, Receive, Scope, Send


# Vendored in Starlette SessionMiddleware from:
# https://github.com/encode/starlette/blob/master/starlette/middleware/sessions.py
# to add the partitioned security flag to enable CHIPS:
# https://developers.google.com/privacy-sandbox/3pcd/chips
# when the session is used within an iframe


class SessionMiddleware:
    def __init__(
        self,
        app: ASGIApp,
        secret_key: str | Secret,
        session_cookie: str = "session",
        max_age: int | None = 14 * 24 * 60 * 60,  # 14 days, in seconds
        path: str = "/",
        same_site: Literal["lax", "strict", "none"] = "lax",
        https_only: bool = False,
    ) -> None:
        self.app = app
        self.signer = itsdangerous.TimestampSigner(str(secret_key))
        self.session_cookie = session_cookie
        self.max_age = max_age
        self.path = path
        self.security_flags = "httponly; samesite=" + same_site
        if https_only or same_site == "none":
            # Secure flag can be used with HTTPS only
            self.security_flags += "; secure"
        if same_site == "none":
            self.security_flags += "; Partitioned;"

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] not in ("http", "websocket"):  # pragma: no cover
            await self.app(scope, receive, send)
            return

        connection = HTTPConnection(scope)
        initial_session_was_empty = True

        if self.session_cookie in connection.cookies:
            data = connection.cookies[self.session_cookie].encode("utf-8")
            try:
                data = self.signer.unsign(data, max_age=self.max_age)
                scope["session"] = json.loads(b64decode(data))
                initial_session_was_empty = False
            except BadSignature:
                scope["session"] = {}
        else:
            scope["session"] = {}

        async def send_wrapper(message: Message) -> None:
            if message["type"] == "http.response.start":
                if scope["session"]:
                    # We have session data to persist.
                    data = b64encode(json.dumps(scope["session"]).encode("utf-8"))
                    data = self.signer.sign(data)
                    headers = MutableHeaders(scope=message)
                    header_value = "{session_cookie}={data}; path={path}; {max_age}{security_flags}".format(  # noqa E501
                        session_cookie=self.session_cookie,
                        data=data.decode("utf-8"),
                        path=self.path,
                        max_age=f"Max-Age={self.max_age}; " if self.max_age else "",
                        security_flags=self.security_flags,
                    )
                    headers.append("Set-Cookie", header_value)
                elif not initial_session_was_empty:
                    # The session has been cleared.
                    headers = MutableHeaders(scope=message)
                    header_value = "{session_cookie}={data}; path={path}; {expires}{security_flags}".format(  # noqa E501
                        session_cookie=self.session_cookie,
                        data="null",
                        path=self.path,
                        expires="expires=Thu, 01 Jan 1970 00:00:00 GMT; ",
                        security_flags=self.security_flags,
                    )
                    headers.append("Set-Cookie", header_value)
            await send(message)

        await self.app(scope, receive, send_wrapper)
```

**Comment by ToasterChicken** [comment]
I imagine there's going to be  lot of specific use cases and differences, I don't think the middleware is intended to take away all of the responsibility for insuring cookies are configured appropriately.  For example there's currently no validating that the connection is over https when setting https_only=True.

Using __Host- or __Secure- is in my opinion trivial, the author sets session_cookie="__Host-session" for example, and make sure they set https_only=True, do you think adding a argument like partitioned=True solve this?  Documentation can be added as appropriate with a link to Mozilla's [Set-Cookie Reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) for guidance.

I already have a pull request ready to submit with a change for #2019 complete with documentation based on @maartenbreddels suggestion with a couple tweaks and some from my original Pull I retracted.  It also updated tests and documentation. Adding a partitioned argument to the pull request would be quick assuming It's not too many changes at once for the maintainers to accept.

So for example the change could include a partitioned argument, 
```python
app.add_middleware(SessionMiddleware, https_only=True, same_site="none",
        partitioned=True, session_cookie="__Host-session", 
        secret_key="ARelatively!LargeComplicatedSecretFor@HMACEfficacy&")
```
Which would result in a cookie:
```
__Host-session=eyJ1c2VyIjogeyJTb21lIjogIkRhdGEifX0=.ZbUnOA.w8LsnA8eqZlA1xvL9J8fmBtCzBI; path=/; Max-Age=1209600; httponly; samesite=none; secure; partitioned
```

Side note for others interested, I suspect this CHIPS change also affected CSP headers as well regarding iFrames and enforcement requiring more verbose policies.
